algorithms/a3c/model.py

```diff
@@ -111,3 +132,94 @@ def forward(self, inputs):
         x = hx
 
         return self.critic_linear(x), self.actor_linear(x), (hx, cx)
+
+
+class A3C_LSTM_GA(torch.nn.Module):
+    """
+    Very similar to the above ActorCritic but has Gated Attention (GA) and processes an instruction
+    which is a part of the input state. The attention enables the policy to focus on certain parts
+    of the input image given the instruction e.g. instruction "Go to the red cup" and a filter
+    could learn and language ground itself in the meaning of "red" with a filter that learns this
+    mapping. There is also a time embedding layer to help stabilize value prediction.
+    Only 3 conv layers compared to ActorCritic's 4 layers.
+    Originally ran on ViZDoom environment with a non-square image state
+    """
+    def __init__(self, num_input_channels, num_outputs, resolution, vocab_size, episode_length):
+        super(A3C_LSTM_GA, self).__init__()
+
+        # Image Processing
+        self.conv1 = nn.Conv2d(num_input_channels, 128, kernel_size=8, stride=4)
+        self.conv2 = nn.Conv2d(128, 64, kernel_size=4, stride=2)
+        self.conv3 = nn.Conv2d(64, 64, kernel_size=4, stride=2)
+
+        self.output_width, self.output_height, \
+            self.num_output_filters = calculate_input_width_height_for_A3C_LSTM_GA(resolution)
+        self.lstm_cell_size = self.output_width * self.output_height * self.num_output_filters
+
+        # Instruction Processing
+        self.gru_hidden_size = 256
+        self.input_size = vocab_size
+        self.embedding = nn.Embedding(self.input_size, 32)
+        self.gru = nn.GRU(32, self.gru_hidden_size)
+
+        # Gated-Attention layers
+        self.attn_linear = nn.Linear(self.gru_hidden_size, 64)
+
+        # Time embedding layer, helps in stabilizing value prediction
```

Comment:
Could you explain more to me about the time embedding layer and why it stabilizes value?

Reply:
I don't actually know "exactly" how but let's work it out here:

if episode is 1000 steps long and the the step number from 0 to 999 get's passed in as tx, and extra self.time_emb_dim dimensions are input to critic+actor to help use time to your advantage (e.g. rush when you're running out of time) or purely just to know that the min reward you can get in 1000 steps with movement reward  -0.001 is -1.0). The agent can work out in sparse reward settings therefore that if there are only 3 steps left it's value prediction will be accurate towards the end and no cup in sight?

How about something like that?

Reply:
i don't see exactly how it can help the actor though? Should you act drastically differently with 30 steps left than with 300? Maybe

Reply:
Yusun helped me figure this out. Value prediction isn't given time input so for the same image state, if there is 1 step left in episode or 900, the value head can't learn to predict less value for the 1st case.